app/controllers/MessageResponseController.scala

```diff
@@ -0,0 +1,86 @@
+
+import controllers.actions.GetArrivalForWriteActionProvider
+import javax.inject.Inject
+import models.ArrivalState
+import models.GoodsReleasedResponse
+import models.MessageResponse
+import models.MessageSender
+import models.UnloadingPermissionResponse
+import play.api.Logger
+import play.api.mvc.Action
+import play.api.mvc.ControllerComponents
+import repositories.ArrivalMovementRepository
+import services.ArrivalMovementService
+import services.XmlValidationService
+import uk.gov.hmrc.play.bootstrap.controller.BackendController
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+import scala.xml.NodeSeq
+
+class MessageResponseController @Inject()(cc: ControllerComponents,
+                                          arrivalMovementService: ArrivalMovementService,
+                                          getArrival: GetArrivalForWriteActionProvider,
+                                          arrivalMovementRepository: ArrivalMovementRepository,
+                                          xmlValidationService: XmlValidationService)(implicit ec: ExecutionContext)
+    extends BackendController(cc) {
+
+  private val logger = Logger(getClass)
+
+  def post(messageSender: MessageSender): Action[NodeSeq] = getArrival(messageSender.arrivalId)(parse.xml).async {
+    implicit request =>
+      val xml: NodeSeq = request.request.body
+
+      val messageResponse: Option[MessageResponse] = request.headers.get("X-Message-Type") match {
+        case Some("IE025") => Some(GoodsReleasedResponse)
+        case Some("IE043") => Some(UnloadingPermissionResponse)
+        case _             => None
+      }
+
+      messageResponse match {
+        case Some(response) =>
+          xmlValidationService.validate(xml.toString, response.xsdFile) match {
+            case Success(_) =>
+              arrivalMovementService.makeMessage(messageSender.messageCorrelationId, response.messageType)(xml) match {
+                case Some(message) =>
+                  val newState: ArrivalState = request.arrival.state.transition(response.messageReceived)
+                  arrivalMovementRepository.addResponseMessage(request.arrival.arrivalId, message, newState).map {
+                    case Success(_) => Ok
+                    case Failure(e) =>
+                      logger.error(s"Failure to add message to movement. Exception: ${e.getMessage}")
+                      InternalServerError
+                  }
+                case None =>
+                  logger.error(s"Failure to parse message")
+                  Future.successful(InternalServerError)
+              }
+            case Failure(e) =>
+              logger.error(s"Failure to validate against XSD. Exception: ${e.getMessage}")
+              Future.successful(BadRequest)
+          }
+        case None =>
+          logger.error(s"Invalid or empty X-Message-Type")
+          Future.successful(InternalServerError)
```

Comment:
Should this be a BadRequest?  invalid or empty X-Message-Type is a failure by the client (EIS/stub) to provide accepted data

Reply:
If X-Message-Type is passed in but not handled, log it out here (thanks Gareth!)